pkg/private/zip/build_zip.py

```diff
@@ -189,24 +189,47 @@ def add_tree(self, tree_top: str, destpath: str, mode: int):
         dest_dir = dest + rel_path_from_top + '/'
       else:
         dest_dir = dest
+      to_write[dest_dir] = None
       for file in files:
         to_write[dest_dir + file] = root + '/' + file
 
     for path in sorted(to_write.keys()):
       content_path = to_write[path]
-      # If mode is unspecified, derive the mode from the file's mode.
-      if mode is None:
-        f_mode = 0o755 if os.access(content_path, os.X_OK) else 0o644
-      else:
-        f_mode = mode
-      entry_info = self.make_zipinfo(path=path, mode=f_mode)
-      entry_info.compress_type = zipfile.ZIP_DEFLATED
-      if not content_path:
-        self.zip_file.writestr(entry_info, '')
-      else:
+      if content_path:
+        # If mode is unspecified, derive the mode from the file's mode.
+        if mode is None:
+          f_mode = 0o755 if os.access(content_path, os.X_OK) else 0o644
+        else:
+          f_mode = mode
+        entry_info = self.make_zipinfo(path=path, mode=f_mode)
+        entry_info.compress_type = zipfile.ZIP_DEFLATED
         with open(content_path, 'rb') as src:
           self.zip_file.writestr(entry_info, src.read())
+      else:
+        # Implicitly created directory
+        dir_path = path
+        if not dir_path.endswith('/'):
+          dir_path += '/'
+        entry_info = self.make_zipinfo(path=dir_path, mode=0o755)
+        entry_info.compress_type = zipfile.ZIP_STORED
+        # Set directory bits
+        entry_info.external_attr |= (UNIX_DIR_BIT << 16) | MSDOS_DIR_BIT
+        self.zip_file.writestr(entry_info, '')
 
+def _load_manifest(args, manifest_fp):
+  manifest_map = {}
+  for entry in json.load(manifest_fp):
+    entry[1] = _combine_paths(args.directory, entry[1])
+    manifest_map[entry[1]] = entry
+  manifest_keys = list(manifest_map.keys())
+  # Add all parent directories of entries that have not been added explicitly.
+  for dest in manifest_keys:
+      parent = dest
+      for _ in range(dest.count("/")):
+        parent, _, _ = parent.rpartition("/")
+        if parent and parent not in manifest_map:
+            manifest_map[parent] = [manifest.ENTRY_IS_DIR, parent, None, "0o755", None, None]
```

Comment:
I think this fails if we had a manifest like
```
foo/bar/baz/bin
abc/def
foo/bar    # an explicit pkg_mkdirs and they wanted something like 711 on it.
```
Perhaps this loop should not add directly to manifest_map. Instead, add to a temp map
and then merge those into manifest_map if they are not in there.

Reply:
I am only adding the directory if it isn't already contained in the updated map (see line 230) and thus only if isn't contained in the original manifest. How is that different from a temp map and merging that?